concent_api/core/payments/sci_callback.py

```diff
@@ -0,0 +1,260 @@
+from core.exceptions import SCICallbackFrameError
+from core.exceptions import SCICallbackPayloadError
+from core.exceptions import SCICallbackPayloadSignatureError
+from core.exceptions import SCICallbackRequestIdError
+from core.exceptions import SCICallbackTimeoutError
+from core.exceptions import SCICallbackTransactionSignatureError
+
+from middleman_protocol.concent_golem_messages.message import SignedTransaction
+from middleman_protocol.concent_golem_messages.message import TransactionRejected
+from middleman_protocol.concent_golem_messages.message import TransactionSigningRequest
+from middleman_protocol.exceptions import MiddlemanProtocolError
+from middleman_protocol.message import ErrorFrame
+from middleman_protocol.message import GolemMessageFrame
+from middleman_protocol.stream import send_over_stream
+from middleman_protocol.stream import unescape_stream
+
+
+class RequestIDGenerator:
+
+    _request_id = 0
+
+    @classmethod
+    def generate_request_id(cls) -> int:
+        cls._request_id += 1
+        return cls._request_id
+
+
+def sci_callback(transaction: Transaction) -> Transaction:
+    """
+    This is callback that submits the transaction to the Signing Service
+    and does not rely on having access to the private key.
+
+    Returns signed Ethereum Transaction.
+
+    The flow is:
+    1. Create a TransactionSigningRequest.
+    2. Establish a TCP connection with Middleman.
+    3. Send the TransactionSigningRequest.
+    4. Read the response and close the connection.
+    5a. If the response is a valid SignedTransaction, copy the signature to the transaction object passed to the
+        callback by SCI and return.
+    5b. Otherwise handle errors by raising an exception and interrupting the callback.
+
+    """
+    assert isinstance(transaction, Transaction)
+
+    # Create a TransactionSigningRequest.
+    transaction_signing_request = create_transaction_signing_request(transaction)
+
+    # Generate request ID.
+    request_id = RequestIDGenerator.generate_request_id()
+
+    # Create a MiddleMan Protocol Golem Message Frame.
+    middleman_message = GolemMessageFrame(
+        payload=transaction_signing_request,
+        request_id=request_id,
+    )
+
+    # Send Frame to MiddleMan through MiddleMan Protocol and receive response.
+    raw_response = send_request_to_middleman(middleman_message)
+
+    # Deserialize received Frame and its payload and handle related errors.
+    signed_transaction = deserialize_response_and_handle_errors(raw_response, request_id)
+
+    assert isinstance(signed_transaction, SignedTransaction)
+
+    # Verify received SignedTransaction signature and handle related errors.
+    verify_data_and_signature(signed_transaction, transaction)
+
+    # If the response is a valid SignedTransaction, copy Ethereum Transaction object into new one and
+    # add SignedTransaction data.
+    return copy_transaction_and_signature(signed_transaction, transaction)
```

Comment:
Okay, this one is a bit misleading - I would suggest changing this function, so it doesn't return anything and would rename it to sth more like:
`sign_ethereum_transaction` or `copy_transaction_signature`, so it would look like:
```python
...
sign_transaction(signed_transaction, ethereum_transaction)
return ethereum_transaction
```

Reply:
Hm but it actually doesn't sign, just copies the signature from one object to another. But I agree that it might not return the value, so I will change that and remove `_and` from name (as it copies only signature now).